target/linux/ath79/dts/qca9558_buffalo_bhr-4grv2.dts

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca9557.dtsi"
+
+/ {
+	compatible = "buffalo,bhr-4grv2", "qca,qca9557";
+	model = "Buffalo BHR-4GRV2";
+
+	aliases {
+		led-status = &power;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+
+		power: power {
+			label = "bhr-4grv2:green:power";
+			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		diag {
+			label = "bhr-4grv2:orange:diag";
+			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		vpn_orange {
+			label = "bhr-4grv2:orange:vpn";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+
+		vpn_green {
+			label = "bhr-4grv2:green:vpn";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
+		};
+
+		eco {
+			label = "eco";
+			gpios = <&gpio 21 GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_0>;
+			linux,input-type = <EV_SW>;
+			debounce-interval = <60>;
+		};
+	};
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x040000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
+				reg = <0x040000 0x010000>;
+				read-only;
+			};
+
+			partition@50000 {
+				label = "rootfs";
+				reg = <0x050000 0xe30000>;
+			};
+
+			partition@e80000 {
+				label = "kernel";
+				reg = <0xe80000 0x170000>;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+
+			firmware@50000 {
```

Comment:
Two nodes on the same level and with the same unit address are considered as mistake in the device tree world:

qca9558_buffalo_bhr-4grv2.dtb: Warning (unique_unit_address): /ahb/spi@1f000000/flash@0/partitions/partition@50000: duplicate unit-address (also used in node /ahb/spi@1f000000/flash@0/partitions/firmware@50000)

Luckly, a patch for creating subpartitions was accepted upstream and is already backported to OpenWrt (2a598bbaa3f75b7051c2453a6ccf706191cf2153).

Please try the following instead:

```
&spi {
	status = "okay";
	num-cs = <1>;

	flash@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <25000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "u-boot";
				reg = <0x000000 0x040000>;
				read-only;
			};

			partition@40000 {
				label = "u-boot-env";
				reg = <0x040000 0x010000>;
				read-only;
			};

			partition@50000 {
				compatible = "fixed-partitions";
				#address-cells = <1>;
				#size-cells = <1>;
				label = "firmware";
				reg = <0x050000 0xfa0000>;

				partition@0 {
					label = "rootfs";
					reg = <0x0 0xe30000>;
				};

				partition@e30000 {
					label = "kernel";
					reg = <0xe30000 0x170000>;
				};
			};

			art: partition@ff0000 {
				label = "art";
				reg = <0xff0000 0x010000>;
				read-only;
			};
		};
	};
};
```

But I guess we have to realise that the underlying issue is something different. Usually we're using a firmware partition in conjunction with a rootfs/kernel splitter for boards where we have kernel + rootfs in consecutively partitions. The reason that a firmware partition exists here is only to reuse the existing (but not really fitting) sysupgrade code.

The ultimate fix would be to use a tar archive as image format similar to the NAND sysupgrade-tar image format. It includes kernel and rootfs as individual files, which can be written to the respective partitions. Parts of such a sysupgrade code is for example implemented for the [ipq40xx Openmesh](https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=target/linux/ipq40xx/base-files/lib/upgrade/openmesh.sh;h=f7aed7c9b6452baeb5fb7de7ee6c0546e814ee1d;hb=HEAD) boards.

The major problem with doing it the right way is that it would break the update from the same board(s) in ar71xx, since the ar71xx image doesn't know how to handle this (new) image format. A restore of the original firmware would be required do start from scratch. If the board supports installation via bootloader (tftp) an additional image might be required, since it is unlikely that the bootloader can handle a tar archive.

Reply:
I sse.
If so, drop "firmware" partition node, can I use this method?
```
	buffalo,bhr-4grv2)
		PART_NAME="rootfs:kernel"
		default_do_upgrade "$ARGV"
		;;
```

Reply:
> If so, drop "firmware" partition node, can I use this method?

To be honest, I wasn't aware that this syntax is used in the tree and I've no idea what it really does. You need to be risky and give it a try.

Reply:
hmm...
OK, I'm compiling the image containing this change now, I will try it.

Reply:
I tried above, and it succeeded!

- ar71xx -> ath79
log: https://gist.github.com/musashino205/a435a20ff98d8bdacf7cf6b2f08578f1

- ath79 -> ath79
log: https://gist.github.com/musashino205/fb21766ddea2300f87e5f3b62228f874

Reply:
wow, great job! Please update the PR accordingly.

Reply:
OK, I'll do it!